File: behavioral/command/task_scheduler/typescript/scheduler.ts
import { Command } from "./commands";

export class TaskScheduler {
  private tasks: Command[] = [];

  public addTask(command: Command): void {
    // Attempt to get a description for logging, default if method doesn't exist
    const description = (command as any).getDescription
      ? (command as any).getDescription()
      : command.constructor.name;
    console.log(`Adding task: ${description}`);
    this.tasks.push(command);
  }

  public async runPendingTasks(): Promise<void> {
    console.log("\n--- Running Scheduled Tasks ---");
    if (this.tasks.length === 0) {
      console.log("No tasks to run.");
      return;
    }

    // Execute tasks sequentially in FIFO order
    // Could also run in parallel with Promise.all if tasks are independent
    for (const task of this.tasks) {
      const description = (task as any).getDescription
        ? (task as any).getDescription()
        : task.constructor.name;
      console.log(`\nExecuting task: ${description}...`);
      try {
        await task.execute(); // Await completion if execute is async
        console.log(`Task ${description} completed.`);
      } catch (error) {
        console.error(`Error executing task ${description}:`, error);
      }
    }

    // Clear the tasks queue after execution
    this.tasks = [];
    console.log("--- All tasks processed ---");
  }

  // Helper for testing or inspection
  public getTaskCount(): number {
    return this.tasks.length;
  }
}

File: behavioral/command/task_scheduler/python/scheduler.py
from typing import List, Deque
from collections import deque
from commands import Command # Relative import within the package

class TaskScheduler:
    """
    The Invoker class. It holds commands and triggers their execution.
    Doesn't know anything about concrete commands or receivers.
    """
    def __init__(self):
        # Using deque for potential efficiency if tasks were added/removed frequently from both ends
        # but a list would also work fine here.
        self._tasks: Deque[Command] = deque()

    def add_task(self, command: Command) -> None:
        """Adds a command (task) to the queue."""
        print(f"Adding task: {command.__class__.__name__}")
        self._tasks.append(command)

    def run_pending_tasks(self) -> None:
        """Executes all tasks currently in the queue."""
        print("\n--- Running Scheduled Tasks ---")
        if not self._tasks:
            print("No tasks to run.")
            return

        # Execute tasks in FIFO order
        while self._tasks:
            task = self._tasks.popleft()
            print(f"\nExecuting task: {task.__class__.__name__}...")
            try:
                task.execute()
                print(f"Task {task.__class__.__name__} completed.")
            except Exception as e:
                print(f"Error executing task {task.__class__.__name__}: {e}")
        print("--- All tasks processed ---")
